static/js/chart/draw.ts

```diff
@@ -88,6 +99,200 @@ function appendLegendElem(
     .attr("href", (d) => d.link || null);
 }
 
+/**
+ * Adds tooltip element within a given container.
+ *
+ * @param containerId
+ */
+function addTooltip(containerId: string): void {
+  d3.select("#" + containerId)
+    .attr("style", "position: relative")
+    .append("div")
+    .attr("id", TOOLTIP_ID)
+    .attr("style", "position: absolute; display: none; z-index: 1");
+}
+
+/**
+ * Position and show the tooltip.
+ *
+ * @param contentHTML innerHTML of the tooltip as a string
+ * @param containerId
+ * @param datapointX
+ * @param datapointY
+ * @param relativeBoundary tooltip boundary relative to its container element
+ */
+function showTooltip(
+  contentHTML: string,
+  containerId: string,
+  datapointX: number,
+  datapointY: number,
+  relativeBoundary: Boundary
+): void {
+  const tooltipSelect = d3
+    .select(`#${containerId}`)
+    .select(`#${TOOLTIP_ID}`)
+    .html(contentHTML);
+  const rect = (tooltipSelect.node() as HTMLDivElement).getBoundingClientRect();
+  const width = rect.width;
+  const height = rect.height;
+  // center tooltip over the datapoint. If this causes the tooltip to overflow the boundary,
+  // place the tooltip against the respective boundary.
+  let left = datapointX - width / 2;
+  if (left < relativeBoundary.left) {
+    left = relativeBoundary.left;
+  } else if (left + width > relativeBoundary.right) {
+    left = relativeBoundary.right - width;
+  }
+  // place the tooltip against the top of the chart area unless there is too little space between it
+  // and the datapoint, then place the tooltip against the bottom of the chart area
+  let top = 0;
+  if (height > datapointY - TOOLTIP_BOTTOM_OFFSET) {
+    top = relativeBoundary.bottom - height;
+  }
+  tooltipSelect.style("left", left + "px").style("top", top + "px");
+}
+
+/**
+ * Gets the timepoint that the mouse is hovering at. Calculation from https://bl.ocks.org/Qizly/8f6ba236b79d9bb03a80.
+ *
+ * @param containerId
+ * @param xScale
+ * @param listOfTimePoints
+ */
+function getHighlightedTime(
+  containerId: string,
+  xScale: d3.ScaleTime<number, number>,
+  listOfTimePoints: number[]
+): number {
+  const mouseX = d3.mouse(
+    d3.select(`#${containerId}`).node() as HTMLElement
+  )[0];
+  const mouseTime = xScale.invert(mouseX).getTime();
+  listOfTimePoints.sort((a, b) => a - b);
+  let idx = d3.bisect(listOfTimePoints, mouseTime);
+  if (idx > 0 && idx < listOfTimePoints.length) {
+    idx =
+      listOfTimePoints[idx] - mouseTime > mouseTime - listOfTimePoints[idx - 1]
+        ? idx - 1
+        : idx;
+  } else if (idx === listOfTimePoints.length) {
+    idx = idx - 1;
+  }
+  return listOfTimePoints[idx];
+}
+
+/**
+ * Adds highlighting and showing a tooltip on hover for a chart.
+ *
+ * @param svg
+ * @param xScale
+ * @param yScale
+ * @param containerId
+ * @param dataGroups
+ * @param colorFn
+ * @param listOfTimePoints
+ * @param highlight
```

Comment:
could you describe these quickly? especially the highlight param. (i realize we haven't been doing this so far...)

also consider making the param name more specific - highlightArea? highlightGroup? highlightPlane?

Reply:
done